File: app/lib/features/reporting/report.dart
/// What a report contains, and what it may never contain (§4.52).
///
/// Three kinds go to one place, because from a caregiver's side there is one
/// question — "this is wrong, or this is missing" — and asking them to sort it
/// into a category first is asking them to do our triage.
///
/// Everything here is a pure function of its arguments. Nothing reads a clock,
/// a device or a database, so a test can state the whole payload and a change
/// to it is a change somebody has to write down.
library;

import 'scrub.dart';

/// The version of this document's shape. The intake rejects one it does not
/// know rather than guessing, and an old build is told to update rather than
/// being quietly half-understood.
const reportSchema = 1;

/// Kept in step with `pubspec.yaml` by `report_payload_test.dart`, which reads
/// the file. Declared rather than looked up so that no plugin sits between the
/// app and its own version number.
const appVersion = '0.3.1';
const appBuild = '5';

enum ReportKind {
  crash('Something crashed', 'A fault the app caught and recovered from'),
  bug('Something is wrong', 'It does not do what it should'),
  idea('Something is missing', 'A word, a setting, a way of working');

  const ReportKind(this.title, this.description);

  final String title;
  final String description;

  String get wire => name;
}

/// What is known about the tablet. No identifier of any kind.
///
/// The model is a class of hardware shared by millions of devices and is the
/// one field that makes a neural voice timing mean anything: "slow" on an iPad
/// mini 5 and "slow" on this year's iPad are different findings.
typedef DeviceFacts = ({
  String platform,
  String osVersion,
  String? model,
  String locale,
});

/// What the person is looking at when they report. Geometry and level, never
/// content.
typedef BoardFacts = ({int rows, int cols, int level, String engine});

/// The finished report, as the intake receives it.
///
/// [detail] is machine-written and is scrubbed here rather than by the caller,
/// so there is no path that reaches this function with an unscrubbed trace and
/// no way to forget.
Map<String, Object?> reportPayload({
  required ReportKind kind,
  required String note,
  required DeviceFacts device,
  required BoardFacts board,
  String? detail,
  Map<String, Object?>? voice,

  /// When the fault happened, for a crash that is being sent later.
  ///
  /// A crash is recorded at the moment it happens and sent on the next healthy
  /// launch, so the two can be a fortnight apart. Without this the arrival time
  /// is the only time anybody has, and a backlog emptying all at once is
  /// indistinguishable from a device failing over and over right now — which is
  /// what §4.67 was, and it took reading object timestamps out of storage to
  /// tell which.
  DateTime? occurredAt,
}) => {
  'schema': reportSchema,
  'kind': kind.wire,
  'app': {'version': appVersion, 'build': appBuild},
  'device': {
    'platform': device.platform,
    'os': device.osVersion,
    'model': ?device.model,
    'locale': device.locale,
  },
  'board': {
    'rows': board.rows,
    'cols': board.cols,
    'level': board.level,
    'engine': board.engine,
  },
  'occurredAt': ?occurredAt?.toUtc().toIso8601String(),
  'note': note.trim(),
  if (detail != null) 'detail': scrubbed(detail),
  'voice': ?voice,
};

/// The most a report may weigh, matching what the intake accepts.
///
/// Enforced here as well so that a report too large is refused with a sentence
/// on the screen rather than by a 413 from somewhere the caregiver cannot see.
const reportSizeLimit = 64 * 1024;

/// The most a caregiver may type. Generous: somebody describing what went
/// wrong should not be counting characters.
const noteLimit = 4000;

/// What the note box is, said beside the note box (§4.79).
///
/// Everything else in a report is cleaned on the way out: the trace loses
/// quoted text and file paths, and `refusalToSend` stops a report still
/// carrying a profile name. That check reads machine-written text only, on
/// purpose — what somebody typed themselves is theirs to type — which leaves
/// this one field reaching us exactly as written.
///
/// So it carries the reason and not just the instruction. "Do not send personal
/// information" is a warning somebody scrolls past; "this is the only part of
/// the report that can carry a name" is a fact they can act on.
const personalInformationNotice =
    'What you type here is sent exactly as written. Everything else in the '
    'report has names and board words taken out first, so this box is the one '
    'part that can carry them. Please leave out names and anything else '
    'personal.';

File: intake/report.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
)

// MaxBody is the largest report accepted, matching reportSizeLimit in the app
// so that a report the app would refuse to send is also one this would refuse
// to take. A caregiver meets the refusal on the screen either way.
const MaxBody = 64 * 1024

// Schema versions this build understands.
//
// An unknown one is refused rather than guessed at, and the app turns that
// refusal into "update the app". Guessing would mean silently filing a report
// whose shape we do not know, which is worse than not taking it.
var knownSchemas = map[int]bool{1: true}

var validKinds = map[string]bool{"crash": true, "bug": true, "idea": true}

// Report is only the part of the document this service reasons about.
//
// Deliberately not the whole payload. The app owns the wire format and will
// add fields to it; a server that unmarshalled into an exhaustive struct would
// have to be redeployed in step with the app, which is the coupling the schema
// number exists to avoid. Everything is kept verbatim in Raw and written to
// storage as it arrived.
type Report struct {
	Schema int    `json:"schema"`
	Kind   string `json:"kind"`
	Note   string `json:"note"`

	App struct {
		Version string `json:"version"`
		Build   string `json:"build"`
	} `json:"app"`

	Device struct {
		Platform string `json:"platform"`
		OS       string `json:"os"`
		Model    string `json:"model"`
		Locale   string `json:"locale"`
	} `json:"device"`

	Board struct {
		Rows   int    `json:"rows"`
		Cols   int    `json:"cols"`
		Level  int    `json:"level"`
		Engine string `json:"engine"`
	} `json:"board"`

	HasDetail bool `json:"-"`
	HasVoice  bool `json:"-"`

	// Raw is exactly the bytes that arrived. What gets stored, so a field this
	// build does not know about is not lost by passing through it.
	Raw json.RawMessage `json:"-"`
}

// ErrBadRequest marks a body this service will not take. Distinguished from a
// storage failure because the two mean opposite things to the app: one says
// "do not try this again", the other says "try again later".
var ErrBadRequest = errors.New("bad request")

func badRequest(format string, args ...any) error {
	return fmt.Errorf("%w: "+format, append([]any{ErrBadRequest}, args...)...)
}

// ParseReport validates a body far enough to know it is a report, and no
// further.
//
// The line it draws: refuse what cannot be filed, accept what can. A note that
// is empty is not this function's business — a caregiver who pressed send on
// an empty note has still told us something by pressing send.
func ParseReport(body []byte) (*Report, error) {
	if len(body) > MaxBody {
		return nil, badRequest("body is %d bytes, over the %d limit", len(body), MaxBody)
	}

	var probe map[string]json.RawMessage
	if err := json.Unmarshal(body, &probe); err != nil {
		return nil, badRequest("not a JSON object: %v", err)
	}

	var r Report
	if err := json.Unmarshal(body, &r); err != nil {
		return nil, badRequest("fields are not the shape expected: %v", err)
	}

	if !knownSchemas[r.Schema] {
		return nil, badRequest("schema %d is not one this build knows", r.Schema)
	}
	if !validKinds[r.Kind] {
		return nil, badRequest("kind %q is not crash, bug or idea", r.Kind)
	}

	_, r.HasDetail = probe["detail"]
	_, r.HasVoice = probe["voice"]
	r.Raw = append(json.RawMessage(nil), body...)

	return &r, nil
}
